drivers/net/ethernet/renesas/rswitch.c

```diff
@@ -2220,40 +2220,48 @@ static enum pf_type rswitch_get_pf_type_by_num(int num)
 
 void rswitch_put_pf(struct l3_ipv4_fwd_param *param)
 {
-	int i, idx, pf_num;
+	int i, idx, pf_used = 0;
 	enum pf_type type;
+	u32 pf_nums[MAX_PF_ENTRIES] = {0};
 
-	rs_write32(RSWITCH_PF_DISABLE_FILTER, param->priv->addr + FWCFCi(param->pf_cascade_index));
-
+	/* First, need to remember used perfect filter nums before cascade filter reset */
 	for (i = 0; i < MAX_PF_ENTRIES; i++) {
-		pf_num = rs_read32(param->priv->addr + FWCFMCij(param->pf_cascade_index, i)) & 0xff;
-		if (!(rs_read32(param->priv->addr + FWCFMCij(param->pf_cascade_index, i)) & RSWITCH_PF_ENABLE_FILTER))
+		u32 mapping_reg = rs_read32(param->priv->addr + FWCFMCij(param->pf_cascade_index, i));
+
+		if (!(mapping_reg & RSWITCH_PF_ENABLE_FILTER)) {
 			break;
+		}
+
+		pf_nums[i] = mapping_reg & 0xff;
+		pf_used++;
+	}
 
-		type = rswitch_get_pf_type_by_num(pf_num);
+	/* Disable and free cascade filter */
+	rs_write32(RSWITCH_PF_DISABLE_FILTER, param->priv->addr + FWCFCi(param->pf_cascade_index));
+	clear_bit(param->pf_cascade_index, param->priv->filters.cascade);
+
+	/* Free all used perfect filters */
+	for (i = 0; i < pf_used; i++) {
```

Comment:
Could we have pf_nums filled unevenly (with holes)? judging by the code above it looks like at can happen and we will skip some filters then.

Reply:
rswitch_setup_pf() initializes it evenly, so such case is not possible. But just to be sure I will change remembering loop